src/mlpack/methods/lmnn/constraints_impl.hpp

```diff
@@ -0,0 +1,328 @@
+                                        arma::mat& outputDistance,
+                                        const arma::mat& dataset,
+                                        const arma::Row<size_t>& labels,
+                                        const size_t begin,
+                                        const size_t batchSize)
+{
+  // Perform pre-calculation. If neccesary.
+  Precalculate(labels);
+
+  arma::mat subDataset = dataset.cols(begin, begin + batchSize - 1);
+  arma::Row<size_t> sublabels = labels.cols(begin, begin + batchSize - 1);
+
+  // KNN instance.
+  KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  // Vectors to store indices.
+  arma::uvec subIndexSame;
+
+  for (size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Calculate impostors.
+    subIndexSame = arma::find(sublabels == uniqueLabels[i]);
+
+    // Perform KNN search with differently labeled points as reference
+    // set and same class points as query set.
+    knn.Train(dataset.cols(indexDiff[i]));
+    knn.Search(subDataset.cols(subIndexSame), k, neighbors, distances);
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexDiff[i].at(neighbors(j));
+
+    // Store impostors.
+    outputNeighbors.cols(begin + subIndexSame) =  neighbors;
+    outputDistance.cols(begin + subIndexSame) =  distances;
+  }
+}
+
+// Generates {data point, target neighbors, impostors} triplets using
+// TargetNeighbors() and Impostors().
+template<typename MetricType>
+void Constraints<MetricType>::Triplets(arma::Mat<size_t>& outputMatrix,
+                                       const arma::mat& dataset,
+                                       const arma::Row<size_t>& labels)
+{
+  // Perform pre-calculation. If neccesary.
+  Precalculate(labels);
+
+  size_t N = dataset.n_cols;
+
+  arma::Mat<size_t> impostors;
+  Impostors(impostors, dataset);
+
+  arma::Mat<size_t> targetNeighbors;
+  TargetNeighbors(targetNeighbors, dataset);
+
+  outputMatrix = arma::Mat<size_t>(3, k * k * N , arma::fill::zeros);
+
+  for (size_t i = 0, r = 0; i < N; i++)
+  {
+    for (size_t j = 0; j < k; j++)
+    {
+      for (size_t l = 0; l < k; l++, r++)
+      {
+        // Generate triplets.
+        outputMatrix(0, r) = i;
+        outputMatrix(1, r) = targetNeighbors(j, i);
+        outputMatrix(2, r) = impostors(l, i);
+      }
+    }
+  }
+}
+
+template<typename MetricType>
+inline void Constraints<MetricType>::Precalculate(
+                                        const arma::Row<size_t>& labels)
+{
+  // Make sure the calculation is necessary.
+  if (precalculated)
+    return;
+
+  uniqueLabels = arma::unique(labels);
+
+  indexSame.resize(uniqueLabels.n_elem);
+  indexDiff.resize(uniqueLabels.n_elem);
+
+  for (size_t i = 0; i < uniqueLabels.n_elem; i++)
+  {
+    // Store same and diff indices.
+    indexSame[i] = arma::find(labels == uniqueLabels[i]);
```

Comment:
I'm wondering if it makes sense to store the result of ``labels == uniqueLabels[i]`` so that it could be reused for the next line

Reply:
Doing so will need something like set substraction or an AND operation like `(labels && !(previousResult))`, which I think will do nothing good here, as it will be more or less the same thing.

Reply:
I think that as we solve #1445 this code will be removed, so I don't think we need to worry about it for now.